projects/gui/src/mainwindow.cpp

```diff
@@ -969,40 +975,108 @@ void MainWindow::copyFen()
 		cb->setText(fen);
 }
 
-void MainWindow::pasteFen()
+PlayerBuilder* MainWindow::findPlayerBuilder(int side, bool useDefault) const
+{
+	const QList<EngineConfiguration> engines
+		= CuteChessApplication::instance()->engineManager()->engines();
+
+	PlayerBuilder * builder = nullptr;
+	if (!useDefault && !m_players[side]->isHuman())
+	{
+		for (const EngineConfiguration config: engines)
+		{
+			if (config.name() == m_players[side]->name())
+			{
+				builder = new EngineBuilder(config);
+				break;
+			}
+		}
+	}
+	if (!builder)
+	{
+		bool ignoreFlag = QSettings()
+				  .value("games/human_can_play_after_timeout", true)
+				  .toBool();
+		builder = new HumanBuilder(CuteChessApplication::userName(),
+						  ignoreFlag);
+	}
+	return builder;
+}
+
+void MainWindow::showPasteErrorDialog(const QString& variant, const QString& text)
+{
+	QMessageBox msgBox(QMessageBox::Critical,
+			    tr("Paste error"),
+			    tr("Invalid FEN or PGN string for the \"%1\" variant:")
+			    .arg(variant),
+			    QMessageBox::Ok, this);
+	msgBox.setInformativeText(text);
+	msgBox.exec();
+
+	return;
+}
+
+void MainWindow::paste()
+{
+	doPaste(true);
+}
+
+void MainWindow::paste2()
+{
+	doPaste(false);
+}
+
+void MainWindow::doPaste(bool useDefault)
 {
 	auto cb = CuteChessApplication::clipboard();
-	if (cb->text().isEmpty())
+	const QString& text = cb->text();
+	if (text.isEmpty())
 		return;
 
 	QString variant = m_game.isNull() || m_game->board() == nullptr ?
 				"standard" : m_game->board()->variant();
 
+	QByteArray buffer = text.toLatin1();
+	PgnStream pgnStream(&buffer, variant);
+	PgnGame pgnGame;
+	bool pgnOk = pgnGame.read(pgnStream);
+
+	if (pgnOk && useDefault)
+		variant = pgnGame.variant();
+
 	auto board = Chess::BoardFactory::create(variant);
-	if (!board->setFenString(cb->text()))
-	{
-		QMessageBox msgBox(QMessageBox::Critical,
-				   tr("FEN error"),
-				   tr("Invalid FEN string for the \"%1\" variant:")
-				   .arg(variant),
-				   QMessageBox::Ok, this);
-		msgBox.setInformativeText(cb->text());
-		msgBox.exec();
+	auto game = new ChessGame(board, new PgnGame());
 
+	pgnOk = pgnOk && game->setMoves(pgnGame);
+	if (!pgnOk &&  !board->setFenString(text))
+	{
 		delete board;
-		return;
+		return showPasteErrorDialog(variant, text);
+	}
+	if(!pgnOk)
+		// action for FEN input
+		game->setStartingFen(text);
+
+	if (useDefault)
+		game->setTimeControl(TimeControl("inf"));
+	else
+	{
+		game->setTimeControl(*m_players[0]->timeControl(), Chess::Side::White);
```

Comment:
Tournament games and their players are deleted automatically when the game ends. So this code line will crash if you run it when the active tab is showing the last game of a finished tournament. This is one reason why I'd like to have copies of `PlayerBuilder` in `MainWindow::tabData`. Or maybe `TournamentPlayer` because it includes time control settings.

Reply:
Thank you.These are very valuable hints. 
Now testing whether `m_player[0].isNull()` or `m_player[1].isNull()`. 
Disabling Menu Item _"Paste FEN/PGN and Use Target Settings"_ if `m_game.isNull()` not only to be safe but to avoid inexpected behaviour for users.